prov/verbs_ud/src/verbs_ud_msg.c

```diff
@@ -0,0 +1,356 @@
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "verbs_ud.h"
+
+static inline ssize_t
+ofi_verbs_ud_recvmsg(struct fid_ep *ep_fid, const struct fi_msg *msg,
+		     uint64_t flags);
+static inline ssize_t
+ofi_verbs_ud_recvv(struct fid_ep *ep_fid, const struct iovec *iov,
+		   void **desc, size_t count,
+		   fi_addr_t src_addr, void *context);
+static inline ssize_t
+ofi_verbs_ud_recv(struct fid_ep *ep_fid, void *buf, size_t len,
+		  void *desc, fi_addr_t src_addr, void *context);
+static ssize_t
+ofi_verbs_ud_sendmsg(struct fid_ep *ep_fid, const struct fi_msg *msg,
+		     uint64_t flags);
+static inline ssize_t
+ofi_verbs_ud_sendv(struct fid_ep *ep_fid, const struct iovec *iov,
+		   void **desc, size_t count, fi_addr_t dest_addr,
+		   void *context);
+static inline ssize_t
+ofi_verbs_ud_senddata(struct fid_ep *ep_fid, const void *buf,
+		      size_t len, void *desc, uint64_t data,
+		      fi_addr_t dest_addr, void *context);
+static inline ssize_t
+ofi_verbs_ud_send(struct fid_ep *ep_fid, const void *buf, size_t len,
+		  void *desc, fi_addr_t dest_addr, void *context);
+
+static inline ssize_t
+ofi_verbs_ud_injectdata(struct fid_ep *ep, const void *buf, size_t len,
+			uint64_t data, fi_addr_t dest_addr);
+
+static inline ssize_t
+ofi_verbs_ud_inject(struct fid_ep *ep, const void *buf,
+		    size_t len, fi_addr_t dest_addr);
+
+static struct fi_ops_msg ofi_verbs_ud_msg_ops = {
+	.size = sizeof(struct fi_ops_msg),
+	.recv = ofi_verbs_ud_recv,
+	.recvv = ofi_verbs_ud_recvv,
+	.recvmsg = ofi_verbs_ud_recvmsg,
+	.send = ofi_verbs_ud_send,
+	.sendv = ofi_verbs_ud_sendv,
+	.sendmsg = ofi_verbs_ud_sendmsg,
+	.inject = ofi_verbs_ud_inject,
+	.senddata = ofi_verbs_ud_senddata,
+	.injectdata = ofi_verbs_ud_injectdata,
+};
```

Comment:
Move this to under the function definitions to avoid the pre-declarations.  These are also used to setup a function pointer table, so inline isn't useful.  The compiler will inline where it makes sense.